File: Datastructures + Algorithms/Datastructures/Fenwick Tree/Fenwick Tree(Binary Indexed Tree)/BIT.cpp
#include<bits/stdc++.h>
using namespace std;

struct bit {

	int n;
	vector<int> b;
	vector<long long> a;

	bit() {}
	bit(int n2, int val = 0) {
		n = n2 + 1;
		a.resize(n);
		b = vector<int>(n, val);
		if (val) {
			for (int i = 0; i < n2; i++) {
				add(i, val);
			}
		}
	}
	bit(vector<int>& b2) {
		n = b2.size() + 1;
		a.resize(n);
		b = b2;
		for (int i = 0; i < b.size(); i++) {
			add(i, b[i]);
		}
	}

	void add(int idx, int val) {
		b[idx] += val;
		for (idx++; idx < n; idx += idx & -idx) {
			a[idx] += val;
		}
	}

	void assign(int idx, int val) {
		int val2 = val;
		val -= b[idx];
		b[idx] = val2;

		add(idx, val);
	}

	long long sum(int idx) {
		long long tot = 0;
		for (idx++; idx > 0; idx -= idx & -idx) {
			tot += a[idx];
		}

		return tot;
	}

	long long sum(int l, int r) {
		return sum(r) - sum(l - 1);
	}
};

File: Datastructures + Algorithms/Datastructures/Fenwick Tree/Fenwick Tree(Binary Indexed Tree)/BIT.py
# all operations are log(n)
class bit:
    def __init__(self, n):
        self.n = n+1
        self.mx_p = 0
        while 1 << self.mx_p < self.n:
            self.mx_p += 1
        self.a = [0]*(self.n)
        self.tot = 0

    def add(self, idx, val):
        self.tot += val
        idx += 1
        while idx < self.n:
            self.a[idx] += val
            idx += idx & -idx

    def sum_prefix(self, idx):
        tot = 0
        idx += 1
        while idx > 0:
            tot += self.a[idx]
            idx -= idx & -idx

        return tot

    def sum(self, l, r):
        return self.sum_prefix(r) - self.sum_prefix(l-1)

    # lowest idx such that sum_prefix(idx) = val else -1 if idx doesn't exist
    # idx is the idx in the underlying array (idx-1)
    def lower(self, val):
        if val > self.tot:
            return -1

        tot = 0
        idx =  0
        for i in range(self.mx_p, -1, -1):
            if idx + (1<<i) < self.n and tot + self.a[idx + (1 << i)] < val:
                tot += self.a[idx + (1 << i)]
                idx += 1 << i

        return idx